<a href="https://colab.research.google.com/github/DATA601-CustSegment-Nov2025-to-Feb2026/New-CustSegment-Project/blob/Andy_test/Customer_Segmentation_(Merging).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Libraries & Load Datasets

In [42]:
# Import necessary libraries
import pandas as pd
import dask.dataframe as dd

from google.colab import drive

In [43]:
# Mount Google Drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [44]:
# Paths to the 8 datasets stored in Google Drive
transaction_file_path = '/content/drive/MyDrive/Customer-Segmentation-files/data/transaction_data.csv'
demographic_file_path = '/content/drive/MyDrive/Customer-Segmentation-files/data/hh_demographic.csv'
product_file_path = '/content/drive/MyDrive/Customer-Segmentation-files/data/product.csv'
causal_data_file_path = '/content/drive/MyDrive/Customer-Segmentation-files/data/causal_data.csv'
coupon_file_path = '/content/drive/MyDrive/Customer-Segmentation-files/data/coupon.csv'
coupon_redemption_file_path = '/content/drive/MyDrive/Customer-Segmentation-files/data/coupon_redempt.csv'
campaign_file_path = '/content/drive/MyDrive/Customer-Segmentation-files/data/campaign_table.csv'
campaign_desc_file_path = '/content/drive/MyDrive/Customer-Segmentation-files/data/campaign_desc.csv'

# Read the 8 datasets from their CSV files
transaction_df = dd.read_csv(
    transaction_file_path,
    dtype={
        'COUPON_DISC': 'float64',
        'COUPON_MATCH_DISC': 'float64'
    }
)

demographic_df = dd.read_csv(
    demographic_file_path,
    dtype={'HOUSEHOLD_SIZE_DESC': 'object'}
)

product_df = dd.read_csv(product_file_path)

causal_df = dd.read_csv(
    causal_data_file_path,
    dtype={'display': 'object'}
)

coupon_df = dd.read_csv(coupon_file_path)

coupon_redemption_df = dd.read_csv(coupon_redemption_file_path)

campaign_df = dd.read_csv(campaign_file_path)

campaign_description_df = dd.read_csv(campaign_desc_file_path)


# Explore Each Dataset: First 5 Rows & Dataset Dimensions

## Transaction Dataset

In [45]:
# Display first 5 rows of the dataset
transaction_df.head(5)

household_key    BASKET_ID  DAY  PRODUCT_ID  QUANTITY  SALES_VALUE  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   
2           2375  26984851472    1     1036325         1         0.99   
3           2375  26984851472    1     1082185         1         1.21   
4           2375  26984851472    1     8160430         1         1.50   

   STORE_ID  RETAIL_DISC  TRANS_TIME  WEEK_NO  COUPON_DISC  COUPON_MATCH_DISC  
0       364        -0.60        1631        1          0.0                0.0  
1       364         0.00        1631        1          0.0                0.0  
2       364        -0.30        1631        1          0.0                0.0  
3       364         0.00        1631        1          0.0                0.0  
4       364        -0.39        1631        1          0.0                0.0

In [46]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_df = transaction_df.shape[0].compute()
num_cols_transaction_df = len(transaction_df.columns)

print("Number of rows:", num_rows_transaction_df)
print("Number of columns:", num_cols_transaction_df)

Number of rows: 2595732
Number of columns: 12


In [47]:
# Convert all column names to lowercase
transaction_df.columns = transaction_df.columns.str.lower()

# Display first 5 rows after the conversion
transaction_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   
2           2375  26984851472    1     1036325         1         0.99   
3           2375  26984851472    1     1082185         1         1.21   
4           2375  26984851472    1     8160430         1         1.50   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  
0       364        -0.60        1631        1          0.0                0.0  
1       364         0.00        1631        1          0.0                0.0  
2       364        -0.30        1631        1          0.0                0.0  
3       364         0.00        1631        1          0.0                0.0  
4       364        -0.39        1631        1          0.0                0.0

## Demographic Dataset

In [48]:
# Display first 5 rows of the dataset
demographic_df.head()

AGE_DESC MARITAL_STATUS_CODE INCOME_DESC HOMEOWNER_DESC      HH_COMP_DESC  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   
2    25-34                   U      25-34K        Unknown     2 Adults Kids   
3    25-34                   U      75-99K      Homeowner     2 Adults Kids   
4    45-54                   B      50-74K      Homeowner     Single Female   

  HOUSEHOLD_SIZE_DESC KID_CATEGORY_DESC  household_key  
0                   2      None/Unknown              1  
1                   2      None/Unknown              7  
2                   3                 1              8  
3                   4                 2             13  
4                   1      None/Unknown             16

In [49]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_demographic_df = demographic_df.shape[0].compute()
num_cols_demographic_df= len(demographic_df.columns)

print("Number of rows:", num_rows_demographic_df)
print("Number of columns:", num_cols_demographic_df)

Number of rows: 801
Number of columns: 8


In [50]:
# Convert all column names to lowercase
demographic_df.columns = demographic_df.columns.str.lower()

# Display first 5 rows after the conversion
demographic_df.head()

age_desc marital_status_code income_desc homeowner_desc      hh_comp_desc  \
0      65+                   A      35-49K      Homeowner  2 Adults No Kids   
1    45-54                   A      50-74K      Homeowner  2 Adults No Kids   
2    25-34                   U      25-34K        Unknown     2 Adults Kids   
3    25-34                   U      75-99K      Homeowner     2 Adults Kids   
4    45-54                   B      50-74K      Homeowner     Single Female   

  household_size_desc kid_category_desc  household_key  
0                   2      None/Unknown              1  
1                   2      None/Unknown              7  
2                   3                 1              8  
3                   4                 2             13  
4                   1      None/Unknown             16

## Product Dataset

In [51]:
# Display first 5 rows of the dataset
product_df.head()

PRODUCT_ID  MANUFACTURER    DEPARTMENT     BRAND            COMMODITY_DESC  \
0       25671             2       GROCERY  National                  FRZN ICE   
1       26081             2  MISC. TRANS.  National  NO COMMODITY DESCRIPTION   
2       26093            69        PASTRY   Private                     BREAD   
3       26190            69       GROCERY   Private      FRUIT - SHELF STABLE   
4       26355            69       GROCERY   Private             COOKIES/CONES   

            SUB_COMMODITY_DESC CURR_SIZE_OF_PRODUCT  
0          ICE - CRUSHED/CUBED                22 LB  
1  NO SUBCOMMODITY DESCRIPTION                       
2         BREAD:ITALIAN/FRENCH                       
3                  APPLE SAUCE                50 OZ  
4            SPECIALTY COOKIES                14 OZ

In [52]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_product_df = product_df.shape[0].compute()
num_cols_product_df = len(product_df.columns)

print("Number of rows:", num_rows_product_df)
print("Number of columns:", num_cols_product_df)

Number of rows: 92353
Number of columns: 7


In [53]:
# Convert all column names to lowercase
product_df.columns = product_df.columns.str.lower()

# Display first 5 rows after the conversion
product_df.head()

product_id  manufacturer    department     brand            commodity_desc  \
0       25671             2       GROCERY  National                  FRZN ICE   
1       26081             2  MISC. TRANS.  National  NO COMMODITY DESCRIPTION   
2       26093            69        PASTRY   Private                     BREAD   
3       26190            69       GROCERY   Private      FRUIT - SHELF STABLE   
4       26355            69       GROCERY   Private             COOKIES/CONES   

            sub_commodity_desc curr_size_of_product  
0          ICE - CRUSHED/CUBED                22 LB  
1  NO SUBCOMMODITY DESCRIPTION                       
2         BREAD:ITALIAN/FRENCH                       
3                  APPLE SAUCE                50 OZ  
4            SPECIALTY COOKIES                14 OZ

## Causal Dataset

In [54]:
# Display first 5 rows of the dataset
causal_df.head()

PRODUCT_ID  STORE_ID  WEEK_NO display mailer
0       26190       286       70       0      A
1       26190       288       70       0      A
2       26190       289       70       0      A
3       26190       292       70       0      A
4       26190       293       70       0      A

In [55]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_causal_df = causal_df.shape[0].compute()
num_cols_causal_df = len(causal_df.columns)

print("Number of rows:", num_rows_causal_df)
print("Number of columns:", num_cols_causal_df)

Number of rows: 36786524
Number of columns: 5


In [56]:
# Convert all column names to lowercase
causal_df.columns = causal_df.columns.str.lower()

# Display first 5 rows after the conversion
causal_df.head()

product_id  store_id  week_no display mailer
0       26190       286       70       0      A
1       26190       288       70       0      A
2       26190       289       70       0      A
3       26190       292       70       0      A
4       26190       293       70       0      A

## Coupon Dataset

In [57]:
# Display first 5 rows of the dataset
coupon_df.head()

COUPON_UPC  PRODUCT_ID  CAMPAIGN
0  10000089061       27160         4
1  10000089064       27754         9
2  10000089073       28897        12
3  51800009050       28919        28
4  52100000076       28929        25

In [58]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_coupon_df = coupon_df.shape[0].compute()
num_cols_coupon_df = len(coupon_df.columns)

print("Number of rows:", num_rows_coupon_df)
print("Number of columns:", num_cols_coupon_df)

Number of rows: 124548
Number of columns: 3


In [59]:
# Convert all column names to lowercase
coupon_df.columns = coupon_df.columns.str.lower()

# Display first 5 rows after the conversion
coupon_df.head()

coupon_upc  product_id  campaign
0  10000089061       27160         4
1  10000089064       27754         9
2  10000089073       28897        12
3  51800009050       28919        28
4  52100000076       28929        25

## Coupon Redemption Dataset

In [60]:
# Display first 5 rows of the dataset
coupon_redemption_df.head()

household_key  DAY   COUPON_UPC  CAMPAIGN
0              1  421  10000085364         8
1              1  421  51700010076         8
2              1  427  54200000033         8
3              1  597  10000085476        18
4              1  597  54200029176        18

In [61]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_coupon_redemption_df = coupon_redemption_df.shape[0].compute()
num_cols_coupon_redemption_df = len(coupon_redemption_df.columns)

print("Number of rows:", num_rows_coupon_redemption_df)
print("Number of columns:", num_cols_coupon_redemption_df)

Number of rows: 2318
Number of columns: 4


In [62]:
# Convert all column names to lowercase
coupon_redemption_df.columns = coupon_redemption_df.columns.str.lower()

# Display first 5 rows after the conversion
coupon_redemption_df.head()

household_key  day   coupon_upc  campaign
0              1  421  10000085364         8
1              1  421  51700010076         8
2              1  427  54200000033         8
3              1  597  10000085476        18
4              1  597  54200029176        18

## Campaign Dataset

In [63]:
# Display first 5 rows of the dataset
campaign_df.head()

DESCRIPTION  household_key  CAMPAIGN
0       TypeA             17        26
1       TypeA             27        26
2       TypeA            212        26
3       TypeA            208        26
4       TypeA            192        26

In [64]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_campaign_df = campaign_df.shape[0].compute()
num_cols_campaign_df = len(campaign_df.columns)

print("Number of rows:", num_rows_campaign_df)
print("Number of columns:", num_cols_campaign_df)

Number of rows: 7208
Number of columns: 3


In [65]:
# Convert all column names to lowercase
campaign_df.columns = campaign_df.columns.str.lower()

# Display first 5 rows after the conversion
campaign_df.head()

description  household_key  campaign
0       TypeA             17        26
1       TypeA             27        26
2       TypeA            212        26
3       TypeA            208        26
4       TypeA            192        26

## Campaign Description Dataset

In [66]:
# Display first 5 rows of the dataset
campaign_description_df.head()

DESCRIPTION  CAMPAIGN  START_DAY  END_DAY
0       TypeB        24        659      719
1       TypeC        15        547      708
2       TypeB        25        659      691
3       TypeC        20        615      685
4       TypeB        23        646      684

In [67]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_campaign_description_df = campaign_description_df.shape[0].compute()
num_cols_campaign_description_df = len(campaign_description_df.columns)

print("Number of rows:", num_rows_campaign_description_df)
print("Number of columns:", num_cols_campaign_description_df)

Number of rows: 30
Number of columns: 4


In [68]:
# Convert all column names to lowercase
campaign_description_df.columns = campaign_description_df.columns.str.lower()

# Display first 5 rows after the conversion
campaign_description_df.head()

description  campaign  start_day  end_day
0       TypeB        24        659      719
1       TypeC        15        547      708
2       TypeB        25        659      691
3       TypeC        20        615      685
4       TypeB        23        646      684

# Joining Dataset

## Join Transaction Dataset With Demographic Dataset

In [69]:
# Join 'transaction_df' with 'demographic_df' on household_key
transaction_demographic_df = transaction_df.merge(demographic_df, on='household_key', how='left')

# Show all columns
pd.set_option('display.max_columns', None)

# Display first 5 rows of the merged dataset
transaction_demographic_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   
2           2375  26984851472    1     1036325         1         0.99   
3           2375  26984851472    1     1082185         1         1.21   
4           2375  26984851472    1     8160430         1         1.50   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0       364        -0.60        1631        1          0.0                0.0   
1       364         0.00        1631        1          0.0                0.0   
2       364        -0.30        1631        1          0.0                0.0   
3       364         0.00        1631        1          0.0                0.0   
4       364        -0.39        1631        1          0.0                0.0   

  age_desc marital_status_code income_desc homeowner_desc hh_comp_desc  \
0     <NA>                <NA>        <NA>           <NA>         <NA>   
1     <NA>                <NA>        <NA>           <NA>         <NA>   
2     <NA>                <NA>        <NA>           <NA>         <NA>   
3     <NA>                <NA>        <NA>           <NA>         <NA>   
4     <NA>                <NA>        <NA>           <NA>         <NA>   

  household_size_desc kid_category_desc  
0                <NA>              <NA>  
1                <NA>              <NA>  
2                <NA>              <NA>  
3                <NA>              <NA>  
4                <NA>              <NA>

In [70]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_df = transaction_demographic_df.shape[0].compute()
num_cols_transaction_demographic_df = len(transaction_demographic_df.columns)

print("Number of rows:", num_rows_transaction_demographic_df)
print("Number of columns:", num_cols_transaction_demographic_df)

Number of rows: 2595732
Number of columns: 19


## Join Merged Dataset (Transaction & Demographic) With Product Dataset

In [71]:
# Join 'transaction_demographic_df' with 'product_df' on product_id
transaction_demographic_product_df = transaction_demographic_df.merge(product_df, on='product_id', how='left')

# Show all columns
pd.set_option('display.max_columns', None)

# Display first 5 rows of the merged dataset
transaction_demographic_product_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0           2375  26984851472    1     1004906         1         1.39   
1           2375  26984851472    1     1033142         1         0.82   
2           2375  26984851472    1     1036325         1         0.99   
3           2375  26984851472    1     1082185         1         1.21   
4           2375  26984851472    1     8160430         1         1.50   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0       364        -0.60        1631        1          0.0                0.0   
1       364         0.00        1631        1          0.0                0.0   
2       364        -0.30        1631        1          0.0                0.0   
3       364         0.00        1631        1          0.0                0.0   
4       364        -0.39        1631        1          0.0                0.0   

  age_desc marital_status_code income_desc homeowner_desc hh_comp_desc  \
0     <NA>                <NA>        <NA>           <NA>         <NA>   
1     <NA>                <NA>        <NA>           <NA>         <NA>   
2     <NA>                <NA>        <NA>           <NA>         <NA>   
3     <NA>                <NA>        <NA>           <NA>         <NA>   
4     <NA>                <NA>        <NA>           <NA>         <NA>   

  household_size_desc kid_category_desc  manufacturer department     brand  \
0                <NA>              <NA>            69    PRODUCE   Private   
1                <NA>              <NA>             2    PRODUCE  National   
2                <NA>              <NA>            69    PRODUCE   Private   
3                <NA>              <NA>             2    PRODUCE  National   
4                <NA>              <NA>            69    PRODUCE   Private   

                commodity_desc          sub_commodity_desc  \
0                     POTATOES  POTATOES RUSSET (BULK&BAG)   
1                       ONIONS     ONIONS SWEET (BULK&BAG)   
2      VEGETABLES - ALL OTHERS                      CELERY   
3               TROPICAL FRUIT                     BANANAS   
4  ORGANICS FRUIT & VEGETABLES             ORGANIC CARROTS   

  curr_size_of_product  
0                 5 LB  
1                40 LB  
2                       
3                40 LB  
4                 1 LB

In [72]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_df = transaction_demographic_product_df.shape[0].compute()
num_cols_transaction_demographic_product_df = len(transaction_demographic_product_df.columns)

print("Number of rows:", num_rows_transaction_demographic_product_df)
print("Number of columns:", num_cols_transaction_demographic_product_df)

Number of rows: 2595732
Number of columns: 25


## Join Merged Dataset (Transaction, Demographic & Product) With Causal Dataset

In [36]:
# Join 'transaction_demographic_product_df' with 'causal_df' on product_id, store_id and week_no
transaction_demographic_product_causal_df = transaction_demographic_product_df.merge(causal_df, on=['product_id', 'store_id', 'week_no'], how='left')

# Persist in memory
transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.persist()

In [37]:
# Show all columns
pd.set_option('display.max_columns', None)

# Display first 5 rows of the merged dataset
transaction_demographic_product_causal_df.head()

household_key    basket_id  day  product_id  quantity  sales_value  \
0            459  32811565816  392     1039682         1         0.99   
1            878  32811571492  392     1084597         1         1.00   
2           2375  32811575940  392      844165         1         1.33   
3            330  32811585245  392      994928         1         0.74   
4            276  32811586367  392      847232         1         1.49   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0     31762         0.00        1553       57          0.0                0.0   
1       297        -0.33        2002       57          0.0                0.0   
2       364        -0.26        1551       57          0.0                0.0   
3       362         0.00        1112       57          0.0                0.0   
4       362         0.00        1912       57          0.0                0.0   

  age_desc marital_status_code income_desc homeowner_desc  hh_comp_desc  \
0     <NA>                <NA>        <NA>           <NA>          <NA>   
1    25-34                   U      35-49K        Unknown  1 Adult Kids   
2     <NA>                <NA>        <NA>           <NA>          <NA>   
3     <NA>                <NA>        <NA>           <NA>          <NA>   
4    35-44                   U      35-49K        Unknown   Single Male   

  household_size_desc kid_category_desc  manufacturer department     brand  \
0                <NA>              <NA>             2    PRODUCE  National   
1                   3                 2            69    GROCERY   Private   
2                <NA>              <NA>           103    GROCERY  National   
3                <NA>              <NA>            69    GROCERY   Private   
4                   1      None/Unknown           869    GROCERY  National   

                   commodity_desc              sub_commodity_desc  \
0                VEGETABLES SALAD                 VARIETY LETTUCE   
1  CONVENIENT BRKFST/WHLSM SNACKS                TOASTER PASTRIES   
2                     SOFT DRINKS  SFT DRNK 2 LITER BTL CARB INCL   
3                            EGGS                   EGGS - MEDIUM   
4                            SOUP                        DRY SOUP   

  curr_size_of_product display mailer  
0                24 CT    <NA>   <NA>  
1              14.6 OZ       0      A  
2                2 LTR    <NA>   <NA>  
3                 1 DZ    <NA>   <NA>  
4           4.2 OZ/2CT    <NA>   <NA>

In [38]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.shape[0].compute()
num_cols_transaction_demographic_product_causal_df = len(transaction_demographic_product_causal_df.columns)

print("Number of rows in transaction_demographic_product_causal_df:", num_rows_transaction_demographic_product_causal_df)
print("Number of columns in transaction_demographic_product_causal_df:", num_cols_transaction_demographic_product_causal_df)

Number of rows in transaction_demographic_product_causal_df: 2596590
Number of columns in transaction_demographic_product_causal_df: 27


In [39]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_df = transaction_demographic_product_df.shape[0].compute()
num_cols_transaction_demographic_product_df = len(transaction_demographic_product_df.columns)

print("Number of rows in transaction_demographic_product_df:", num_rows_transaction_demographic_product_df)
print("Number of columns in transaction_demographic_product_df:", num_cols_transaction_demographic_product_df)

Number of rows in transaction_demographic_product_df: 2595732
Number of columns in transaction_demographic_product_df: 25


In [40]:
# Calculate the number of row difference between 'transaction_demographic_product_causal_df' and 'transaction_demographic_product_df'
row_difference_1 = num_rows_transaction_demographic_product_causal_df - num_rows_transaction_demographic_product_df

print(f"Row difference between 'transaction_demographic_product_causal_df' and 'transaction_demographic_product_df': {row_difference_1} rows")

Row difference between 'transaction_demographic_product_causal_df' and 'transaction_demographic_product_df': 858 rows


In [41]:
# List all columns except 'display' and 'mailer'
cols_to_check = [col for col in transaction_demographic_product_causal_df.columns
                 if col not in ['display', 'mailer']]

# Create a boolean column marking duplicates within each partition
def mark_duplicates(df, subset):
    return df[df.duplicated(subset=subset, keep=False)]

# Apply to each partition
duplicates_transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.map_partitions(
    mark_duplicates, subset=cols_to_check
)

# Compute number of duplicate rows
num_duplicates_transaction_demographic_product_causal_df  = len(duplicates_transaction_demographic_product_causal_df.compute())

print(f"Number of duplicate rows: {num_duplicates_transaction_demographic_product_causal_df}")

Number of duplicate rows: 1716


In [73]:
# Display first 5 rows of the merged dataset
duplicates_transaction_demographic_product_causal_df.head(5)

# After the left join, the the number of rows increased from 2,595,732 to 2,596,590 (858 rows).
# This is due to multiple matching rows for 'display' and 'mailer' columns in the right table (causal_df).

household_key    basket_id  day  product_id  quantity  sales_value  \
2773           1305  32895350599  399      879948         4          1.0   
2774           1305  32895350599  399      879948         4          1.0   
4990             17  32989921691  405      994928         6          2.0   
4991             17  32989921691  405      994928         6          2.0   
5818           1208  33015899361  407      994928         3          1.0   

      store_id  retail_disc  trans_time  week_no  coupon_disc  \
2773       329        -1.38        1012       58          0.0   
2774       329        -1.38        1012       58          0.0   
4990       361        -2.44        1356       59          0.0   
4991       361        -2.44        1356       59          0.0   
5818       361        -1.22        2254       59          0.0   

      coupon_match_disc age_desc marital_status_code income_desc  \
2773                0.0    19-24                   B   Under 15K   
2774                0.0    19-24                   B   Under 15K   
4990                0.0      65+                   B   Under 15K   
4991                0.0      65+                   B   Under 15K   
5818                0.0     <NA>                <NA>        <NA>   

     homeowner_desc      hh_comp_desc household_size_desc kid_category_desc  \
2773         Renter      1 Adult Kids                   2                 1   
2774         Renter      1 Adult Kids                   2                 1   
4990      Homeowner  2 Adults No Kids                   2      None/Unknown   
4991      Homeowner  2 Adults No Kids                   2      None/Unknown   
5818           <NA>              <NA>                <NA>              <NA>   

      manufacturer department     brand     commodity_desc  \
2773           531    DRUG GM  National  CANDY - CHECKLANE   
2774           531    DRUG GM  National  CANDY - CHECKLANE   
4990            69    GROCERY   Private               EGGS   
4991            69    GROCERY   Private               EGGS   
5818            69    GROCERY   Private               EGGS   

                  sub_commodity_desc curr_size_of_product display mailer  
2773  CANDY BARS (SINGLES)(INCLUDING              1.55 OZ       0      A  
2774  CANDY BARS (SINGLES)(INCLUDING              1.55 OZ       1      A  
4990                   EGGS - MEDIUM                 1 DZ       0      D  
4991                   EGGS - MEDIUM                 1 DZ       6      D  
5818                   EGGS - MEDIUM                 1 DZ       0      D

In [74]:
# Count how many duplicate rows have display equal to "0" in 'duplicates_transaction_demographic_product_causal_df'
display_equal_0 = duplicates_transaction_demographic_product_causal_df[
    duplicates_transaction_demographic_product_causal_df['display'] == "0"
].shape[0].compute()

print("Number of duplicate rows have display equal to '0' in 'duplicates_transaction_demographic_product_causal_df':", display_equal_0, 'rows')

Number of duplicate rows have display equal to '0' in 'duplicates_transaction_demographic_product_causal_df': 858 rows


In [75]:
# As our aim is to focus on transactional-level patterns for customer segmentation, the extra rows don't represent new purchases.
# They exist because a single transaction matched multiple display locations (the same product purchased in the same transaction may have different display locations.)
# Keeping these duplicates could inflate the number of transactions and bias our customer segmnetation, making it look like some customers buy more than they actually do.

# Remove all duplicate rows where display column equals to "0" (0 means not on display) in transaction_demographic_product_causal_df
# Convert Dask DataFrame to pandas
transaction_demographic_product_causal_df_pandas = transaction_demographic_product_causal_df.compute()

# Define columns to check for duplicates (excluding 'display' and 'mailer')
cols_to_identify_transaction_demographic_product_causal_df_pandas = [col for col in transaction_demographic_product_causal_df_pandas.columns if col not in ['display', 'mailer']]

# Remove duplicate rows where display column equals to "0" (0 means not on display) in transaction_demographic_product_causal_df
transaction_demographic_product_causal_df_pandas = transaction_demographic_product_causal_df_pandas[~(
    (transaction_demographic_product_causal_df_pandas.duplicated(subset=cols_to_identify_transaction_demographic_product_causal_df_pandas, keep=False)) &
    (transaction_demographic_product_causal_df_pandas['display'] == "0")
)]

# Convert back to Dask DataFrame
transaction_demographic_product_causal_df = dd.from_pandas(transaction_demographic_product_causal_df_pandas)

# Persist in memory
transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.persist()

In [76]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.shape[0].compute()
num_cols_transaction_demographic_product_causal_df = len(transaction_demographic_product_causal_df.columns)

print("Number of rows in transaction_demographic_product_causal_df:", num_rows_transaction_demographic_product_causal_df)
print("Number of columns in transaction_demographic_product_causal_df:", num_cols_transaction_demographic_product_causal_df)

Number of rows in transaction_demographic_product_causal_df: 2595732
Number of columns in transaction_demographic_product_causal_df: 27


## Join Coupon Dataset With Coupon Redemption Dataset

In [77]:
# Join 'coupon_df' with 'coupon_redemption_df' on coupon_upc and campaign
merged_coupon_df = coupon_df.merge(coupon_redemption_df, on=['coupon_upc', 'campaign'], how='left')

# Persist in memory
merged_coupon_df = merged_coupon_df.persist()

In [78]:
# Show all columns
pd.set_option('display.max_columns', None)

# Display first 5 rows of the merged dataset
merged_coupon_df.head(5)

coupon_upc  product_id  campaign  household_key    day
0  10000089061       27160         4            NaN    NaN
1  10000089064       27754         9          321.0  446.0
2  10000089064       27754         9         1773.0  439.0
3  10000089073       28897        12         1229.0  491.0
4  10000089073       28897        12         1633.0  497.0

In [79]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_merged_coupon_df = merged_coupon_df.shape[0].compute()
num_cols_merged_coupon_df = len(merged_coupon_df.columns)

print("Number of rows:", num_rows_merged_coupon_df)
print("Number of columns:", num_cols_merged_coupon_df)

Number of rows: 2218551
Number of columns: 5


**Interpretation for merged_coupon_df:**

**Row 0:**
- Coupon 10000089061 for product 27160 in campaign 4.
- household_key = NaN, day = NaN → *coupon exists but has not been redeemed yet.*

**Row 1:**
- Coupon 10000089064 for product 27754 in campaign 9.
- household_key = 321, day = 446 → *household 321 redeemed this coupon on day 446.*

**Row 2:**
- Same coupon and campaign as row 1.
- household_key = 1773, day = 439 → *same coupon redeemed by another household 1773 on day 439.*

## Join Merged Dataset (Transaction, Demographic, Product & Causal) With Merged Coupon Dataset

In [80]:
# Join 'transaction_demographic_product_causal_df' with 'merged_coupon_df' on household_key, day and product_id.
merged_transaction_coupon_df = transaction_demographic_product_causal_df.merge(merged_coupon_df, on=['household_key', 'day', 'product_id'], how='left')

# Persist in memory
merged_transaction_coupon_df = merged_transaction_coupon_df.persist()

In [81]:
# Show all columns
pd.set_option('display.max_columns', None)

# Display first 5 rows of the merged dataset
merged_transaction_coupon_df.head(5)

household_key    basket_id  day  product_id  quantity  sales_value  \
0            608  32811570135  392     1096573         1         3.99   
1            608  32811570135  392    13039578         1         3.49   
2            878  32811571492  392     1084208         1         0.80   
3           1453  32811562421  392      986912         1         1.99   
4           1453  32811562421  392     1139830         1         0.39   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0       297        -0.50         958       57          0.0                0.0   
1       297         0.00         958       57          0.0                0.0   
2       297        -0.29        2002       57          0.0                0.0   
3     31862         0.00        2059       57          0.0                0.0   
4     31862         0.00        2059       57          0.0                0.0   

  age_desc marital_status_code income_desc homeowner_desc   hh_comp_desc  \
0    25-34                   B      25-34K         Renter   1 Adult Kids   
1    25-34                   B      25-34K         Renter   1 Adult Kids   
2    25-34                   U      35-49K        Unknown   1 Adult Kids   
3    45-54                   A    125-149K      Homeowner  2 Adults Kids   
4    45-54                   A    125-149K      Homeowner  2 Adults Kids   

  household_size_desc kid_category_desc  manufacturer department     brand  \
0                  5+                3+           608    GROCERY  National   
1                  5+                3+           794    GROCERY  National   
2                   3                 2          2135    GROCERY  National   
3                   3                 1          4166       DELI  National   
4                   3                 1           678     PASTRY  National   

                   commodity_desc              sub_commodity_desc  \
0          FRZN MEAT/MEAT DINNERS               SNACKS/APPETIZERS   
1  CONVENIENT BRKFST/WHLSM SNACKS                     CEREAL BARS   
2                   CANNED JUICES  SV BEV: BEV/JUICE 50-100% JUIC   
3                   PREPARED FOOD      PREP FD: SIDE DISHES (HOT)   
4                           BREAD          BREAD:SAVORY/SPECIALTY   

  curr_size_of_product display mailer  coupon_upc  campaign  
0                16 OZ    <NA>   <NA>         NaN       NaN  
1              4.87 OZ    <NA>   <NA>         NaN       NaN  
2                16 OZ    <NA>   <NA>         NaN       NaN  
3                         <NA>   <NA>         NaN       NaN  
4                 2 OZ    <NA>   <NA>         NaN       NaN

In [82]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_merged_transaction_coupon_df = merged_transaction_coupon_df.shape[0].compute()
num_cols_merged_transaction_coupon_df = len(merged_transaction_coupon_df.columns)

print("Number of rows in merged_transaction_coupon_df:", num_rows_merged_transaction_coupon_df)
print("Number of columns in merged_transaction_coupon_df:", num_cols_merged_transaction_coupon_df)

Number of rows in merged_transaction_coupon_df: 2595914
Number of columns in merged_transaction_coupon_df: 29


In [83]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_transaction_demographic_product_causal_df = transaction_demographic_product_causal_df.shape[0].compute()
num_cols_transaction_demographic_product_causal_df = len(transaction_demographic_product_causal_df.columns)

print("Number of rows in transaction_demographic_product_causal_df:", num_rows_transaction_demographic_product_causal_df)
print("Number of columns in transaction_demographic_product_causal_df:", num_cols_transaction_demographic_product_causal_df)

Number of rows in transaction_demographic_product_causal_df: 2595732
Number of columns in transaction_demographic_product_causal_df: 27


In [84]:
# Calculate the number of row difference between 'transaction_demographic_product_causal_df' and 'merged_transaction_coupon_df'
row_difference_2 = num_rows_merged_transaction_coupon_df - num_rows_transaction_demographic_product_causal_df

print(f"Row difference between 'transaction_demographic_product_causal_df' and 'merged_transaction_coupon_df': {row_difference_2} rows")

Row difference between 'transaction_demographic_product_causal_df' and 'merged_transaction_coupon_df': 182 rows


In [85]:
# List all columns except 'display' and 'mailer'
cols_to_check = [col for col in merged_transaction_coupon_df.columns
                 if col not in ['coupon_upc', 'campaign', 'household_key']]

# Create a boolean column marking duplicates within each partition
def mark_duplicates(df, subset):
    return df[df.duplicated(subset=subset, keep=False)]

# Apply to each partition
duplicates_merged_transaction_coupon_df = merged_transaction_coupon_df.map_partitions(
    mark_duplicates, subset=cols_to_check
)

# Compute number of duplicate rows
num_duplicates_merged_transaction_coupon_df = len(duplicates_merged_transaction_coupon_df.compute())

print(f"Number of duplicate rows: {num_duplicates_merged_transaction_coupon_df}")

Number of duplicate rows: 313


In [86]:
# Convert 'coupon_upc' column from gloat to string to avoid scientific notation caused by large numeric ID.
duplicates_merged_transaction_coupon_df['coupon_upc'] = duplicates_merged_transaction_coupon_df['coupon_upc'].astype('str')

# Display first 5 rows of the merged dataset
duplicates_merged_transaction_coupon_df.head(5)

# After the left join, the the numner of rows increased from 2,596,590 to 2,596,722 (182 rows).
# This is due to multiple matching rows for 'coupon_upc' and 'campaign' columns in the right table (merged_transaction_coupon_df).

household_key    basket_id  day  product_id  quantity  sales_value  \
93292            573  33118502623  415      849697         1         2.99   
93293            573  33118502623  415      849697         1         2.99   
96439           1337  33132154250  416      995876         1         2.19   
96440           1337  33132154250  416      995876         1         2.19   
98360           1098  33132674533  416      920710         1         3.74   

       store_id  retail_disc  trans_time  week_no  coupon_disc  \
93292       445        -0.50        2152       60         -1.0   
93293       445        -0.50        2152       60         -1.0   
96439       364        -0.20        1920       60         -1.0   
96440       364        -0.20        1920       60         -1.0   
98360       343        -1.25        1849       60          0.0   

       coupon_match_disc age_desc marital_status_code income_desc  \
93292                0.0     <NA>                <NA>        <NA>   
93293                0.0     <NA>                <NA>        <NA>   
96439                0.0    45-54                   U      50-74K   
96440                0.0    45-54                   U      50-74K   
98360                0.0     <NA>                <NA>        <NA>   

      homeowner_desc   hh_comp_desc household_size_desc kid_category_desc  \
93292           <NA>           <NA>                <NA>              <NA>   
93293           <NA>           <NA>                <NA>              <NA>   
96439      Homeowner  2 Adults Kids                   4                 2   
96440      Homeowner  2 Adults Kids                   4                 2   
98360           <NA>           <NA>                <NA>              <NA>   

       manufacturer department     brand       commodity_desc  \
93292          1225    GROCERY  National  FLUID MILK PRODUCTS   
93293          1225    GROCERY  National  FLUID MILK PRODUCTS   
96439          1485    GROCERY  National   FROZEN BREAD/DOUGH   
96440          1485    GROCERY  National   FROZEN BREAD/DOUGH   
98360          4971    DRUG GM  National   FIRST AID PRODUCTS   

                   sub_commodity_desc curr_size_of_product display mailer  \
93292    REFRIGERATED COFFEE CREAMERS                32 OZ    <NA>   <NA>   
93293    REFRIGERATED COFFEE CREAMERS                32 OZ    <NA>   <NA>   
96439  FRZN GARLIC BREAD/TOAST/STICKS             11.25 OZ       9      0   
96440  FRZN GARLIC BREAD/TOAST/STICKS             11.25 OZ       9      0   
98360           FIRST AID MEDICATIONS                         <NA>   <NA>   

          coupon_upc  campaign  
93292  10000085364.0       8.0  
93293  55000013033.0       8.0  
96439  57045970076.0       8.0  
96440  10000085361.0       8.0  
98360  50008130078.0       8.0

In [87]:
# As our aim is to focus on transactional-level patterns for customer segmentation, the extra rows don't represent new purchases.
# They exist because a single transaction matched multiple coupons or campaigns (a customer may have used or been eligible for multiple coupons or campaigns on the same product purchase.)
# Keeping these duplicates could inflate the number of transactions and bias our customer segmnetation, making it look like some customers buy more than they actually do.

# Define columns to check for duplicates (excluding 'coupon_upc' and 'campaign')
cols_to_identify_merged_transaction_coupon_df = [
    'household_key', 'basket_id', 'day', 'product_id', 'quantity',
    'sales_value', 'store_id', 'retail_disc', 'trans_time', 'week_no', 'coupon_disc',
    'coupon_match_disc', 'age_desc', 'marital_status_code', 'income_desc', 'homeowner_desc',
    'hh_comp_desc', 'household_size_desc', 'kid_category_desc',
    'manufacturer', 'department', 'brand', 'commodity_desc',
    'sub_commodity_desc', 'curr_size_of_product', 'display', 'mailer'
]

# We want to keep the coupon with the highest dicount
# Sort by discount descending and keep first per transaction
merged_transaction_coupon_df = merged_transaction_coupon_df.sort_values('coupon_disc', ascending=True)
merged_transaction_coupon_df_no_dup = merged_transaction_coupon_df.drop_duplicates(subset=cols_to_identify_merged_transaction_coupon_df , keep='first')

# Persist in memory
merged_transaction_coupon_df_no_dup = merged_transaction_coupon_df_no_dup.persist()

# Display first 5 rows of the merged dataset
merged_transaction_coupon_df_no_dup.head(5)

household_key    basket_id  day  product_id  quantity  sales_value  \
2253151           1228  31553560338  308    10118257         8        57.52   
41479              679  32957147025  403     5978656         0         0.00   
41498              679  32957147063  403     5978656         0         0.00   
291897            1901  33919382569  465    13511457        24        60.00   
943153             400  41124522239  627     5978656         0         0.00   

         store_id  retail_disc  trans_time  week_no  coupon_disc  \
2253151       327        -6.40        1113       45       -55.93   
41479         447         0.00        1826       58       -37.93   
41498         447         0.00        1837       58       -36.50   
291897        319       -52.56        1613       67       -34.00   
943153        388         0.00        1508       90       -31.46   

         coupon_match_disc age_desc marital_status_code income_desc  \
2253151                0.0    45-54                   U    100-124K   
41479                  0.0     <NA>                <NA>        <NA>   
41498                  0.0     <NA>                <NA>        <NA>   
291897                 0.0    45-54                   U      35-49K   
943153                 0.0    35-44                   A    150-174K   

        homeowner_desc   hh_comp_desc household_size_desc kid_category_desc  \
2253151        Unknown  Single Female                   1      None/Unknown   
41479             <NA>           <NA>                <NA>              <NA>   
41498             <NA>           <NA>                <NA>              <NA>   
291897       Homeowner  2 Adults Kids                   3                 1   
943153       Homeowner  2 Adults Kids                   3                 1   

         manufacturer  department     brand commodity_desc  \
2253151            69  MEAT-PCKGD   Private    MEAT - MISC   
41479               1              National                  
41498               1              National                  
291897           1208     GROCERY  National    SOFT DRINKS   
943153              1              National                  

                     sub_commodity_desc curr_size_of_product display mailer  \
2253151             GRND/PATTY - FROZEN                  2LB       9      0   
41479                                                           <NA>   <NA>   
41498                                                           <NA>   <NA>   
291897   SOFT DRINKS 12/18&15PK CAN CAR                12 OZ       7      D   
943153                                                          <NA>   <NA>   

         coupon_upc  campaign  
2253151         NaN       NaN  
41479           NaN       NaN  
41498           NaN       NaN  
291897          NaN       NaN  
943153          NaN       NaN

In [88]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_merged_transaction_coupon_df_no_dup = merged_transaction_coupon_df_no_dup.shape[0].compute()
num_cols_merged_transaction_coupon_df_no_dup = len(merged_transaction_coupon_df_no_dup.columns)

print("Number of rows in merged_transaction_coupon_df:", num_rows_merged_transaction_coupon_df_no_dup)
print("Number of columns in merged_transaction_coupon_df:", num_cols_merged_transaction_coupon_df_no_dup)

Number of rows in merged_transaction_coupon_df: 2595732
Number of columns in merged_transaction_coupon_df: 29


## Join Merged Transaction & Coupon Dataset With Campaign Description Dataset

In [89]:
# Join 'merged_transaction_coupon_df_no_dup' with 'campaign_description_df' on campaign
merged_transaction_coupon_campaign_df = merged_transaction_coupon_df_no_dup.merge(campaign_description_df, on=['campaign'], how='left')
final_merged_df = merged_transaction_coupon_campaign_df

# Persist in memory
final_merged_df = final_merged_df.persist()

In [90]:
# Calculate the number of rows and columns in the Dask dataFrame
num_rows_final_merged_df = final_merged_df.shape[0].compute()
num_cols_final_merged_df = len(final_merged_df.columns)

print("Number of rows in merged_transaction_coupon_df:", num_rows_final_merged_df)
print("Number of columns in merged_transaction_coupon_df:", num_cols_final_merged_df)

Number of rows in merged_transaction_coupon_df: 2595732
Number of columns in merged_transaction_coupon_df: 32


In [91]:
final_merged_df.head(5)

household_key    basket_id  day  product_id  quantity  sales_value  \
0           1228  31553560338  308    10118257         8        57.52   
1            679  32957147025  403     5978656         0         0.00   
2            679  32957147063  403     5978656         0         0.00   
3           1901  33919382569  465    13511457        24        60.00   
4            400  41124522239  627     5978656         0         0.00   

   store_id  retail_disc  trans_time  week_no  coupon_disc  coupon_match_disc  \
0       327        -6.40        1113       45       -55.93                0.0   
1       447         0.00        1826       58       -37.93                0.0   
2       447         0.00        1837       58       -36.50                0.0   
3       319       -52.56        1613       67       -34.00                0.0   
4       388         0.00        1508       90       -31.46                0.0   

  age_desc marital_status_code income_desc homeowner_desc   hh_comp_desc  \
0    45-54                   U    100-124K        Unknown  Single Female   
1     <NA>                <NA>        <NA>           <NA>           <NA>   
2     <NA>                <NA>        <NA>           <NA>           <NA>   
3    45-54                   U      35-49K      Homeowner  2 Adults Kids   
4    35-44                   A    150-174K      Homeowner  2 Adults Kids   

  household_size_desc kid_category_desc  manufacturer  department     brand  \
0                   1      None/Unknown            69  MEAT-PCKGD   Private   
1                <NA>              <NA>             1              National   
2                <NA>              <NA>             1              National   
3                   3                 1          1208     GROCERY  National   
4                   3                 1             1              National   

  commodity_desc              sub_commodity_desc curr_size_of_product display  \
0    MEAT - MISC             GRND/PATTY - FROZEN                  2LB       9   
1                                                                        <NA>   
2                                                                        <NA>   
3    SOFT DRINKS  SOFT DRINKS 12/18&15PK CAN CAR                12 OZ       7   
4                                                                        <NA>   

  mailer  coupon_upc  campaign description  start_day  end_day  
0      0         NaN       NaN        <NA>        NaN      NaN  
1   <NA>         NaN       NaN        <NA>        NaN      NaN  
2   <NA>         NaN       NaN        <NA>        NaN      NaN  
3      D         NaN       NaN        <NA>        NaN      NaN  
4   <NA>         NaN       NaN        <NA>        NaN      NaN

In [92]:
# Save 'final_merged_df' as CSV file
final_merged_df.compute().to_csv('customer_segmentation_after_merging.csv', index=False)